front/src/services/bedrock.ts

```diff
@@ -0,0 +1,175 @@
+				if (result.status === "rejected") {
+					console.error(`Failed to upload file: ${files[index].name}`, result.reason);
+					toast.error(`Failed to upload file: ${files[index].name}`);
+					return false;
+				}
+				return true;
+			})
+			.map((result) => result.value);
+	}
+
+	// The difference between this method and the uploadFiles method is that this method saves the file hash to blockchain so that it can be fetched later
+	async saveFiles(...fileInfos: Omit<FileFullInfos, "post_hash">[]): Promise<FileEntry[]> {
+		fileInfos = z.array(FileFullInfosSchema.omit({ post_hash: true })).parse(fileInfos); // Validate the input because fileEntry can be built without using the parse method
+
+		const fileEntries = (
+			await Promise.allSettled(
+				fileInfos.map(async ({ path, ...rest }) => ({ post_hash: await this.postFile(rest), path })),
+			)
+		)
+			.filter((result): result is PromiseFulfilledResult<FileEntry> => result.status === "fulfilled")
+			.map((result) => result.value);
+		let newFiles: FileEntry[] = [];
+		await this.alephService.updateAggregate(FILE_ENTRIES_AGGREGATE_KEY, EncryptedFileEntriesSchema, ({ files }) => {
+			const decryptedFiles = this.decryptFilesPaths(files);
+			newFiles = fileEntries.filter(({ path }) => {
+				if (this.fileExists(decryptedFiles, path)) {
+					console.error(`File already exists at path: ${path}`);
+					toast.error(`File already exists at path: ${path}`);
+					return false;
+				}
+				return true;
+			});
+
+			return {
+				files: [
+					...files,
+					...newFiles.map(({ post_hash, path }) => ({
+						post_hash,
+						path: EncryptionService.encryptEcies(path, this.alephService.encryptionPrivateKey.publicKey.compressed),
+					})),
+				],
+			};
+		});
+		if (newFiles.length === 0) return Promise.reject("No new files were saved");
+		return newFiles;
+	}
+
+	private async postFile({ key, iv, ipfs_hash }: Omit<FileFullInfos, "post_hash" | "path">): Promise<string> {
+		const { item_hash } = await this.alephService.createPost(FILE_POST_TYPE, {
+			key: EncryptionService.encryptEcies(key, this.alephService.encryptionPrivateKey.publicKey.compressed),
+			iv: EncryptionService.encryptEcies(iv, this.alephService.encryptionPrivateKey.publicKey.compressed),
+			ipfs_hash,
+		});
+		return item_hash;
+	}
+
+	fileExists(files: FileEntry[], path: string): boolean {
+		return files.some((entry) => entry.path === path);
+	}
+
+	private decryptFilesPaths(files: EncryptedFileEntry[]): FileEntry[] {
+		return files.map(({ post_hash, path }) => ({
+			post_hash,
+			path: EncryptionService.decryptEcies(path, this.alephService.encryptionPrivateKey.secret),
+		}));
+	}
+
+	async fetchFileEntries(): Promise<FileEntry[]> {
+		return (await this.alephService.fetchAggregate(FILE_ENTRIES_AGGREGATE_KEY, EncryptedFileEntriesSchema)).files.map(
+			({ post_hash, path }) => ({
+				post_hash,
+				path: EncryptionService.decryptEcies(path, this.alephService.encryptionPrivateKey.secret),
+			}),
+		);
+	}
+}
```

Comment:
ok, looks very complex for a pretty simple feature but we'll see later if needed to refactor it